### PR TITLE
Required fields and fragments

### PR DESCRIPTION
**Pull request by stubailo**
TODO:

- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests pass
- [x] Update CHANGELOG.md with your change
- [x] If this was a change that affects the external API, update the README

**Pull Request Labels**

- [x] feature
- [x] blocking
- [ ] docs

### CODE REVIEW

**Comment by stubailo** [comment]
cc @stevehollaar and @jnwng please give this a look if you have time! My intention is to make this rule a reliable way to ensure that a certain field is fetched (`id` in our case). @stevehollaar took the first step by being more conservative with named fragments, and I covered a few more cases with certain combinations of inline fragments.

Open to any ideas that allow more valid combinations of fields, but I'd also rather err on the conservative side.

**Comment by stubailo** [comment]
cc @hellendag

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Change log
 ### vNEXT
+- BREAKING: The `required-fields` rule has been significantly changed to make it a completely reliable method of ensuring `id` fields are always requested when available. [PR #199](https://github.com/apollographql/eslint-plugin-graphql/pull/199) Here is the behavior, let's say we are requiring field `x`:
+    - On any field whose return type defines a field called `x`, the selection set must directly contain `x`.
+    - In any named fragment declaration whose type defines a field called `x`, the selection set must directly contain `x`.
+    - An inline fragment whose type defines a field called `x` must contain `x` in its selection set unless its parent is also an inline fragment that contains the field `x`.
+    - Here's a specific case which is _no longer valid_:
+        - `query { greetings { hello ... on Greetings { id } } }`
+        - This must now be written as `query { greetings { id hello ... on Greetings { id } } }`
+    - This is a more convervative approach than before, driven by the fact that it's quite hard to ensure that a combination of inline fragments actually covers all of the possible types of a selection set.
```
Comment (holic) [inline CHANGELOG.md]: nit: typo at "convervative"?

File: test/validationRules/required-fields.js [hunk 2]
```diff
@@ -31,14 +32,42 @@ const requiredFieldsTestCases = {
...
+    {
+      code:
+        "const x = gql`query { greetingOrStory { ... on Greetings { id } ... on Story { comments { text } } } }`",
+      errors: [
+        {
+          message: `'id' field required on '... on Story'`,
+          type: "TaggedTemplateExpression"
+        }
+      ]
+    },
```
Comment (hellendag) [inline test/validationRules/required-fields.js]: Also worth testing an interface where `id` is a shared field?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,13 @@
 # Change log
 ### vNEXT
+- BREAKING: The `required-fields` rule has been significantly changed to make it a completely reliable method of ensuring `id` fields are always requested when available. [PR #199](https://github.com/apollographql/eslint-plugin-graphql/pull/199) Here is the behavior, let's say we are requiring field `x`:
+    - On any field whose return type defines a field called `x`, the selection set must directly contain `x`.
```
Comment (hellendag) [inline CHANGELOG.md]: `x` is a bit of an odd example field name. :D

File: src/customGraphQLValidationRules.js [hunk 4]
```diff
@@ -1,60 +1,131 @@
...
+                }'`,
+                [node]
+              )
+            );
+          }
         }
+      });
+    },
+
+    // Every inline fragment must have an ID specified inside itself or in some
```
Comment (hellendag) [inline src/customGraphQLValidationRules.js]: s/`an ID`/`the required field`

File: test/validationRules/required-fields.js (hunk 2 above)
Comment (stubailo) [reply test/validationRules/required-fields.js]: Done!

**Comment by stubailo** [comment]
(Waiting on an OK from @jnwng about whether this breaking change is fine)

**Pull request merged**